src/lang/eval/StateIPCClient.ml

```diff
@@ -0,0 +1,194 @@
+  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+  A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+ 
+  You should have received a copy of the GNU General Public License along with
+  scilla.  If not, see <http://www.gnu.org/licenses/>.
+*)
+open Core
+open Result.Let_syntax
+open MonadUtil
+open Syntax
+open JSON
+open ParserUtil
+open TypeUtil
+open StateIPCIdl
+open ErrorUtils
+
+module ER = ParserRep
+
+module IPCClient = IPCIdl(Idl.GenClient ())
+
+(* Translate JRPC result to our result. *)
+let translate_res res =
+  match res with
+  | Error (e : RPCError.err_t) ->
+    fail0 (Printf.sprintf "StateIPCClient: Error in IPC access: (code:%d, message:%s)." e.code e.message)
+  | Ok res' -> pure res'
+
+let ipcclient_exn_wrapper thunk =
+  try
+    thunk()
+  with
+  | Unix.Unix_error (_, s1, s2) ->
+    fail0 ("StateIPCClient: Unix error: " ^ s1 ^ s2)
+  | _ -> fail0 "StateIPCClient: Unexpected error making JSON-RPC call"
+
+(* Send msg via socket s with a delimiting character "0xA". *)
+let send_delimited oc msg =
+  let msg' = msg ^ "\n" in
+  Out_channel.output_string oc msg';
+  Out_channel.flush oc
+
+let binary_rpc ~socket_addr (call: Rpc.call) : Rpc.response =
+  let socket = Unix.socket ~domain: Unix.PF_UNIX ~kind: Unix.SOCK_STREAM ~protocol:0 in
+  Unix.connect socket ~addr:(Unix.ADDR_UNIX socket_addr);
+  let (ic, oc) = Unix.in_channel_of_descr socket, Unix.out_channel_of_descr socket in
+  let msg_buf = Jsonrpc.string_of_call ~version: Jsonrpc.V2 call in
+  DebugMessage.plog (Printf.sprintf "Sending: %s\n" msg_buf);
+  (* Send data to the socket. *)
+  let _ = send_delimited oc msg_buf in
+  (* Get response. *)
+  let response = Caml.input_line ic in
+  Unix.close socket;
+  DebugMessage.plog (Printf.sprintf "Response: %s\n" response);
+  Jsonrpc.response_of_string response
+
+(* Encode a literal into bytes, opaque to the backend storage. *)
+let serialize_literal l = Bytes.of_string (PrettyPrinters.literal_to_jstring l)
+let deserialize_literal s tp =
+  try
+    pure @@ ContractState.jstring_to_literal s tp
+  with
+  | Invalid_json s ->
+    fail (s @ mk_error0 "StateIPCClient: Error deserializing literal fetched from IPC call")
+
+(* Map fields are serialized into Ipcmessage_types.MVal
+   Other fields are serialized using serialize_literal into bytes/string. *)
+let rec serialize_field value =
+  match value with
+  | Map (_, mlit) ->
+    let mpb = Caml.Hashtbl.fold (fun key value acc ->
+      let key' = Bytes.to_string (serialize_literal key) in
+      (* values can be Maps or non-map literals. Hence a recursive call. *)
+      let val' = serialize_field value in
+      (key', val') :: acc
+    ) mlit [] in
+    Ipcmessage_types.Mval({ Ipcmessage_types.m = mpb })
+    (* If there are maps _inside_ a non-map field, they are treated same
+     * as non-Map field values are not serialized as protobuf maps. *)
+  | _ -> Ipcmessage_types.Bval (serialize_literal value)
```

Comment:
If we have a map inside a non-map structure, such as `Option {Map(String)(String)}`, then such a map is always treated as a full value. It is not encoded into the protobuf object as a map (since we can't do that), but the whole `Option` object is serialized as a string and treated as bytes by protobuf. (The map itself is serialized as our legacy JSON object - the blockchain cannot and doesn't care).